controller/backup_target_controller.go

```diff
@@ -377,6 +444,55 @@ func (btc *BackupTargetController) reconcile(name string) (err error) {
 	return nil
 }
 
+func (btc *BackupTargetController) setDefaultBackupTarget(backupTarget *longhorn.BackupTarget) (*longhorn.BackupTarget, error) {
```

Comment:
I think we can change the logic to the following 
so it is more easy to understand
```
log := getLoggerForBackupTarget(btc.logger, backupTarget)
backupTargetMap, err := btc.ds.ListBackupTargets()
if err != nil {
	log.WithError(err).Errorf("Failed to get all backup targets")
	return nil, err
}

if len(backupTargetMap) != 1 && !backupTarget.Spec.Default {
        // if current backup target is not the only one and it is not set as default
	backupTarget.Status.Default = false
	backupTarget, err = btc.ds.UpdateBackupTargetStatus(backupTarget)
	if err != nil {
		log.WithError(err).Errorf("Failed to update backup target status")
		return nil, err
	}
	return backupTarget, nil
}

// len(backupTargetMap) == 1 or backupTarget.Spec.Default == true will enter this block
// always update others to non-default
// and always update self status/spec default to True
for _, bt := range backupTargetMap {
	if bt.Name != backupTarget.Name && bt.Status.Default {
		bt.Status.Default = false
		bt, err := btc.ds.UpdateBackupTargetStatus(bt)
		if err != nil {
			log.WithError(err).Errorf("Failed to update backup target %s status", bt.Name)
			return nil, err
		}
		bt.Spec.Default = false
		if _, err := btc.ds.UpdateBackupTarget(bt); err != nil {
			log.WithError(err).Errorf("Failed to update backup target %s spec", bt.Name)
			return nil, err
		}
	}
}

backupTarget.Spec.Default = true
backupTarget, err = btc.ds.UpdateBackupTarget(backupTarget)
if err != nil {
	log.WithError(err).Errorf("Failed to update backup target spec")
	return nil, err
}
backupTarget.Status.Default = true
backupTarget, err = btc.ds.UpdateBackupTargetStatus(backupTarget)
if err != nil {
	log.WithError(err).Errorf("Failed to update backup target status")
	return nil, err
}

return backupTarget, nil

```

Reply:
Though "User can not delete or disable a default backup target directly."
I sill add the logic of updating the backuptarget status default to false in the `if{}`
You can remove it to further simplify it and return directly if we are pretty sure that won't happen

Reply:
Sure, I would like to add this logic `updating the backup target status default to false in the if{}`
Thanks, and I will have a update ASAP.

Reply:
I am also thinking, should we update the current BackupTarget spec/status at this moment?
Or can we just assign the value
and update the BackupTarget in the end of the reconcile in defer function with deep equal like other controllers?

Forgot to mention in the previous comment(in my code), we should check if we really need to update the spec/status to reduce the overhead. Thanks

Reply:
because if there will be a race condition when other controllers try to get the correct  default backuptarget
even you update the spec/status immediately there is still a chance that other controller get right before you update it.

Reply:
There will be extra checks and copies of the backup target to do if we update `Status.Default=True` at the end of the reconciliation. 
The `Status.Default=True` might not be updated because it copies the backup target as an existing backup target (with Status.Default=True) and compares it with the backup target  (with Status.Default=True) that synchronizes the status (Status might not be changed) at the end of the reconciliation.

Reply:
Decided not to move the `setDefaultBackupTarget `